tanjun/parsing.py

```diff
@@ -31,8 +31,457 @@
+
+        if not all(n.startswith("-") for n in names):
+            raise ValueError("All option names must start with `-`")
+
+        if flags and GREEDY in flags:
+            raise ValueError("Option cannot be greedy")
+
+        self.empty_value = empty_value
+        self.names = names
+        super().__init__(key, converters=converters, default=default, flags=flags)
+
+    def __repr__(self) -> str:
+        return f"{type(self).__name__} <{self.key}, {self.names}>"
+
+
+class ShlexParser(traits.Parser):
+    __slots__: typing.Sequence[str] = ("_arguments", "_options")
+
+    def __init__(self, *, parameters: typing.Optional[typing.Iterable[traits.Parameter]] = None) -> None:
+        self._arguments: typing.MutableSequence[traits.Argument] = []
+        self._options: typing.MutableSequence[traits.Option] = []
+
+        if parameters is not None:
+            self.set_parameters(parameters)
+
+    @property
+    def parameters(self) -> typing.Sequence[traits.Parameter]:
+        return (*self._arguments, *self._options)
+
+    def add_parameter(self, parameter_: traits.Parameter, /) -> None:
+        if isinstance(parameter_, traits.Option):
+            self._options.append(parameter_)
+
+        else:
+            self._arguments.append(parameter_)
+            found_final_argument = False
+
+            for argument_ in self._arguments:
+                if found_final_argument:
+                    del self._arguments[-1]
+                    raise ValueError("Multi or greedy argument must be the last argument")
+
+                found_final_argument = MULTI in argument_.flags or GREEDY in argument_.flags
+
+    def remove_parameter(self, parameter_: traits.Parameter, /) -> None:
+        if isinstance(parameter_, traits.Option):
+            self._options.remove(parameter_)
+
+        else:
+            self._arguments.remove(parameter_)
+
+    def set_parameters(self, parameters: typing.Iterable[traits.Parameter], /) -> None:
+        self._arguments = []
+        self._options = []
+
+        for parameter_ in parameters:
+            self.add_parameter(parameter_)
+
+    async def parse(
+        self, ctx: traits.Context, /
+    ) -> typing.Tuple[typing.Sequence[typing.Any], typing.Mapping[str, typing.Any]]:
+        parser = SemanticShlex(ctx)
+        arguments = await parser.get_arguments(self._arguments)
+        options = await parser.get_options(self._options)
+        return arguments, options
+
+
+def generate_parameters(command: _CommandT, /, *, ignore_self: bool = True) -> _CommandT:
+    # TODO: implement this to enable generating parameters from a function's signature.
+    if command.parser is None:
+        raise RuntimeError("Cannot generate parameters for a command with no parser")
+
+    if command.function is None:
+        raise RuntimeError("Cannot generate parameters for a command with no function")
+
+    raise NotImplementedError
+
+
+def verify_parameters(command: _CommandT, /) -> _CommandT:
```

Comment:
Might want to have this and a type aware version

Reply:
Actually this is a shitty idea, trying to verify the types would be a nightmare.